src/core/net/tcp6.hpp

```diff
@@ -311,6 +310,36 @@ class Tcp : public InstanceLocator, private NonCopyable
          * @retval kErrorFailed  Failed to deinitialize the TCP endpoint.
          */
         Error Deinitialize(void);
+
+    private:
+        enum : uint8_t
+        {
+            kNumTimers = 4,
+        };
+
+        static_assert(sizeof(mTimers[0]) >= sizeof(TimerMilli), "mTimers entries are too small");
+        static_assert(OT_ARRAY_LENGTH(mTimers) == kNumTimers, "mTimers has an incorrect array length");
+
+        TimerMilli &GetTimer(size_t aIndex) { return *reinterpret_cast<TimerMilli *>(&mTimers[aIndex].mSize[0]); }
+
+        static Endpoint &FromTimer(TimerMilli &aTimer, size_t aIndex)
+        {
+            uint8_t *ptr = reinterpret_cast<uint8_t *>(&aTimer);
+            ptr -= offsetof(Endpoint, mTimers[aIndex].mSize[0]);
+            return *reinterpret_cast<Endpoint *>(ptr);
+        }
```

Comment:
I agree with the drawbacks you've identified. We had a discussion earlier about the question of having OpenThread maintain a pool, and we had decided not to do that. The deciding argument being that maintaining a pool in OpenThread would require OpenThread to be re-compiled if the pool size is changed, but that's undesirable in certain situations.

I like the idea you suggested below, of having a single timer in the TCP class. My understanding is that, on expiry, it would iterate over all TCP endpoints and check the four timers on each one to check which ones expired. And when a timer expires or is cancelled, it would iterate over all TCP endpoints to find the timer that expires next. That would be inefficient when there is a large number of TCP endpoints, but it will probably OK since we only expect there to be few TCP endpoints.

Now, we could improve on that (decreasing computation time but increasing memory cost) by having intrusive doubly-linked lists in each TCP endpoint. That way, we can maintain a sorted list for each timer, and avoid having to iterate over the whole linked list. But this is getting into the territory of what `Timer` is support to handle, and it's a bit disappointing that `Timer` is proving unsuitable for this. I'm wondering whether the best thing to do would be to change `Timer` to make it suitable --- for example, remove the inheritance from `InstanceLocator` and provide a C struct counterpart to include in `otTcpEndpoint`. What are your thoughts on doing this?

Reply:
> I agree with the drawbacks you've identified. We had a discussion earlier about the question of having OpenThread maintain a pool, and we had decided not to do that. The deciding argument being that maintaining a pool in OpenThread would require OpenThread to be re-compiled if the pool size is changed, but that's undesirable in certain situations.

I know we discussed this and I do see that you already spent the time to find a solution/code to make it work but I'd still think it will be good for us to re-visit it.

I think different solutions may address all situations, and longer term having `TcpEndPoint` be an internal definition (i.e., not having it as `otTcpEndPoint` in public header) will help simplify things.

The key drawback for me (in addition to complexity it adds) is that in this model, we may end up over-estimating the size (use larger than necessary buffers for `otTcpEndPoint`) which is not a good for memory-contained devices.

I don't want us to penalize the memory-constrained case in favor of flexibility for the not-constrained case.

Here is my view on this:

- On a memory-constrained device, one way or another, the app developer needs to determine the the max number of TCP end points that may be needed. I think this would apply independent of whether it is being managed by app layer code or we use a pool and have an OT configuration to specify the pool size.
   - And as I said there are already other items (like max number of unicast/multicast address) that have a hard max limit which is determined by an  `OPENTRHEAD_CONFIG` and if the app need changes, we need to re-build OT stack.

- If we need flexibility and support for dynamic number of end-points (which I guess makes sense on a not-very-memory constrained platform), then
  1) We can still use a pool with large enough estimate for number of entries that cover the application use-cases.
      - Even if the pool size needs a change, requiring a re-build of the OT stack should not add a lot of overhead.
  2) We can have an option (under certain `OPENTHREAD_CONFIG`) where OT stack uses heap to allocate the `TcpEndPoint` instance instead of having a pool.
      - This way it is more dynamic and tied to platform (heap size available)
      - For example, this makes sense for POSIX like platform where we can have way more memory available and can use heap.
   3) We can also consider adding a third option (again under some `OPENTHREAD_CONFIG`) where we still require the user to provide the buffer for the TCP end points but instead of defining the `otTcpEndPoint` struct in public API, we have an API where the user can ask and get the size of the buffer needed for a TCP end-point object. Then we expect the caller to provide a buffer of at least that size to us (which we then use `new` to convert to a `TcpEndPoint` instance).

Thoughts?

Reply:
I don't mind revisiting this issue. I don't think it would be much work to change the API to use a pool-based approach, if that's what we decide to do. It would be great if @jwhui and @simonlingoogle could share their thoughts as well, as they participated in previous discussions of this point.

There are a few things I'd like to point out though:
* I understand the argument of possibly wasting memory (4 B per timer). But more memory is wasted simply by using a `Timer` object for each timer in a TCP endpoint. This is because (1) each timer has its own `mNext` pointer, (2) each timer has its own `mHandler` pointer, and (3) each timer potentially has a pointer to its enclosing otInstance. That's 8-12 bytes of overhead per timer. Adding 0-4 bytes due to overestimating the size of a TCP timer seems small in comparison. If we want to really save RAM, we should take the approach of having a single global timer, and multiplexing all the TCP timers on top of it. That way, we can share the same linked list for all timers (sorted in order of each TCP endpoint's earliest timer). In my opinion, this is the best solution.
* In my view, the root cause of the problem is not exposing `otTcpEndpoint` in C. Rather, it is the fact that we can't put a `Timer` in a C struct. I think it's reasonable to expect this to work because other RTOSes I've used (TinyOS and RIOT OS) allow it. So, finding a way to expose timers to C applications is probably the best way to address the root cause of this (though it would still be less efficient than what's proposed above). All we would need to do to provide a version of `TimerMilli` that doesn't inherit from `InstanceLocator`.

I think there are advantages to having applications allocate `otTcpEndpoint`s directly. I also think that there are advantages to having a pool of them allocated by OpenThread. However, I don't think that the issue we're seeing here with `Timer`s is a good reason to switch, for the two reasons I mentioned above.

Thoughts? @abtink @simonlingoogle @jwhui

Reply:
> the root cause of the problem is not exposing otTcpEndpoint in C. Rather, it is the fact that we can't put a Timer in a C struct.

This is now happening with `Timer` but can happen with other internal types and definitions. By having a C public `otTcpEndpoint` definition, we basically limit what we can include in it. Whereas if it is opaque public type and we define it  `class TcpEndPoint` as internal C++ type, we can use all other types/definitions from OT in it. For example,  we can have it inherit from `InstanceLocator` (which then optimizes how `mInstance` is stored), or use `Tasklet`, `LinkedList<>` or even other helper types (say `RssAverager`).

In terms of design, I think if the user does not need to know/access certain info, it's better/safer to hide the info. We don't intend the user of TCP API to access `otTcpEndPoint` fields anyway, so not having a definition of this struct in  in public C header makes it a simpler model. The only thing we really care is for user to know the size (if we expect user to provide buffer for its instances which we can do from an API).

> That's 8-12 bytes of overhead per timer. Adding 0-4 bytes due to overestimating the size of a TCP timer seems small in comparison. 

Yes! That's the reason I suggested remembering the expire time and sharing a `Timer` (insead of multple timers)

>  and multiplexing all the TCP timers on top of it. That way, we can share the same linked list for all timers (sorted in order of each TCP endpoint's earliest timer)

This is actually somewhat similar to what `Timer` implementation does!... :) 

> I think it's reasonable to expect this to work because other RTOSes I've used (TinyOS and RIOT OS) allow it. So, finding a way to expose timers to C applications is probably the best way to address the root cause of this (though it would still be less efficient than what's proposed above)

The fact that we don't expose `Timer` or `Tasklet` (or other OS/platform related functions) in OT public APIs is very much intentional. We do not want OpenThread to act as or replace an OS.  OT is intended as Thread stack implementation which is portable and can run on different OSes and platforms.

Reply:
Thanks @abtink for clarifying this. I hadn't previously realized that it was recommended to avoid using a large number of `Timer`s, though in retrospect, this makes total sense.

I had realized that sharing a linked list for all timers is how `Timer` and `TimerScheduler` worked; the reason I had originally tried to instantiate `Timer`s within the TCP endpoints was to avoid reimplementing this. But based on what you've said above, I understand why creating that many `Timer`s is a bad idea.

Here's a concrete proposal for moving forward with this. I'll rewrite the timer code to have a single `Timer` in the TCP class. This seems to be the best option since we both agree that this is the most memory-efficient.

Meanwhile, we'll leave the question as to whether to have a pool of `otTcpEndpoint`s open for additional discussion; hopefully @jwhui and @simonlingoogle will chime in. I don't want merging in the TCP implementation to stall while we discuss what to do about the `otTcpEndpoint` API, since it sounds like both can proceed independently for now.

Does that sound like a good plan?

Reply:
Sounds good to me. Thanks. 👍

Reply:
> Meanwhile, we'll leave the question as to whether to have a pool of `otTcpEndpoint`s open for additional discussion; hopefully @jwhui and @simonlingoogle will chime in. I don't want merging in the TCP implementation to stall while we discuss what to do about the `otTcpEndpoint` API, since it sounds like both can proceed independently for now.

OpenThread core hardcoding the size of the pool requires recompiling the library whenever the pool size needs to change. One concern with recompiling is that it does not allow Thread Group's certification-by-inheritance, as the component-level certification would no longer applied to a recompiled library.

Reply:
How about this model from https://github.com/openthread/openthread/pull/6790#discussion_r683057291?

>  we still require the user to provide the buffer for the TCP end points but instead of defining the `otTcpEndPoint` struct in public API, we have an API where the user can ask and get the size of the buffer needed for a TCP end-point object. Then we expect the caller to provide a buffer of at least that size to us (which we then use new to convert to a TcpEndPoint instance).

It lets us hide the definition in `TcpEndPoint` while being flexble and requireing the user to manage the buffer/space for them. The only downside is it puts the burdon on caller to get the size from API call and provide buffer of at least that size (if OT is built as a library and linked with different app layer code, the returned value from API (i.e. the size) would be fixed/constant).

Reply:
A pool TCP sockets of works for me. 

#6036 defined the opaque TCP structure to address this problem, but it's tricky to get the size right.

```
struct otTcpSocket
{
#if __SIZEOF_POINTER__ == 8
    uint64_t mMemHolder[35];
#elif __SIZEOF_POINTER__ == 4
    uint32_t mMemHolder[50];
#else
#error "only 32bit and 64bit are supported"
#endif
};
```